File: app/src/main/java/com/example/mlallemant/mentalbattle/UI/Menu/Fragment/CreateJoinFragment.java
package com.example.mlallemant.mentalbattle.UI.Menu.Fragment;

import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;
import androidx.fragment.app.FragmentTransaction;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.ProgressBar;
import android.widget.TextView;
import android.widget.Toast;

import com.example.mlallemant.mentalbattle.R;
import com.example.mlallemant.mentalbattle.UI.Menu.MenuActivity;
import com.example.mlallemant.mentalbattle.Utils.DatabaseManager;
import com.example.mlallemant.mentalbattle.Utils.Player;
import com.example.mlallemant.mentalbattle.Utils.Session;
import com.example.mlallemant.mentalbattle.Utils.Utils;

import static com.facebook.FacebookSdk.getApplicationContext;

/**
 * Created by m.lallemant on 15/11/2017.
 */

public class CreateJoinFragment extends Fragment {

    //UI
    private TextView tv_title;
    private ImageView iv_back;
    private EditText et_session_name;
    private EditText et_session_password;
    private Button btn_create_session;
    private ProgressBar pb_create_session;

    //Utils
    private Player currentPlayer;
    private DatabaseManager db;
    private boolean isCreator = false;


    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View v = inflater.inflate(R.layout.menu_create_fragment, container, false);

        MenuActivity menuActivity = (MenuActivity) getActivity();
        currentPlayer = menuActivity.getCurrentPlayer();

        db = DatabaseManager.getInstance();
        db.deletePlayerInLobby(currentPlayer);

        isCreator = getArguments().getBoolean("creator");

        initUI(v);
        initListener();

        return v;
    }


    private void initUI(View v){
        iv_back = (ImageView) v.findViewById(R.id.iv_create_back);
        et_session_name = (EditText) v.findViewById(R.id.et_create_session_name);
        et_session_password = (EditText) v.findViewById(R.id.et_create_session_password);
        btn_create_session = (Button) v.findViewById(R.id.btn_create_session);
        tv_title = (TextView) v.findViewById(R.id.tv_title_session);
        pb_create_session = (ProgressBar) v.findViewById(R.id.pb_create_session);

        pb_create_session.setVisibility(View.INVISIBLE);
        if (isCreator){
            btn_create_session.setText("CREATE");
            tv_title.setText("Create Session");
        } else {
            btn_create_session.setText("JOIN");
            tv_title.setText("Join Session");
        }
    }

    private void initListener(){
        iv_back.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                returnSelectorFragment();
                db.insertPlayerInLobby(currentPlayer);
            }
        });

        btn_create_session.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                pb_create_session.setVisibility(View.VISIBLE);
                if (isCreator){
                    createSession();
                } else {
                    joinSession();
                }

            }
        });

    }

    private void createSession() {
        String sessionName = et_session_name.getText().toString();
        String sessionPassword = et_session_password.getText().toString();

        if (sessionName.length() > 4 && sessionPassword.length() > 5) {

            final Session session = new Session(sessionName, sessionPassword, Utils.SESSION_STATE_WAITING);
            db.initCheckSessionExist(session);

            db.setOnSessionExistListener(new DatabaseManager.OnSessionExistListener() {
                @Override
                public void notifyUser(boolean isExist) {
                    if (isExist){
                        makeToast("Session already exists");
                    } else {
                        db.insertSession(session);
                        Player playerForSession = new Player(currentPlayer.getId(), currentPlayer.getName(), currentPlayer.getScore(), Utils.SESSION_RDY_YES, Utils.SESSION_CREATOR);
                        db.insertPlayerInSession(session, playerForSession);
                        launchSessionLauncherFragment(true);
                    }
                    pb_create_session.setVisibility(View.INVISIBLE);
                }
            });

        } else {
            makeToast("name or password not enough long");
        }

    }

    private void joinSession(){
        String sessionName = et_session_name.getText().toString();
        String sessionPassword = et_session_password.getText().toString();


        if (sessionName.length() > 0 && sessionPassword.length() > 0) {
            final Session session = new Session(sessionName, sessionPassword, Utils.SESSION_STATE_WAITING);
            db.initCheckSessionExist(session);

            db.setOnSessionExistListener(new DatabaseManager.OnSessionExistListener() {
                @Override
                public void notifyUser(boolean isExist) {
                    if (isExist){
                        Player playerForSession = new Player(currentPlayer.getId(), currentPlayer.getName(), currentPlayer.getScore(), Utils.SESSION_RDY_KO, Utils.SESSION_NEW_YES);
                        db.insertPlayerInSession(session, playerForSession);
                        launchSessionLauncherFragment(false);
                    } else {
                        makeToast("Bad name or password");
                    }
                    pb_create_session.setVisibility(View.INVISIBLE);
                }
            });
        }else {
            makeToast("Bad name or password");
        }


    }

    private void makeToast(String text){
        Toast.makeText(getApplicationContext(), text,
                Toast.LENGTH_SHORT).show();
    }


    private void returnSelectorFragment(){
        SelectorFragment selectorFragment = new SelectorFragment();
        FragmentManager fm = getFragmentManager();
        FragmentTransaction ft = fm.beginTransaction();
        ft.replace(R.id.menu_fl_select, selectorFragment);
        ft.commit();
    }

    private void launchSessionLauncherFragment(boolean isCreator){

        SessionLauncherFragment sessionLauncherFragment = new SessionLauncherFragment();
        Bundle bundle = new Bundle();
        bundle.putBoolean("creator", isCreator);
        bundle.putString("name", et_session_name.getText().toString());
        bundle.putString("password", et_session_password.getText().toString());
        sessionLauncherFragment.setArguments(bundle);
        FragmentManager fm = getFragmentManager();
        FragmentTransaction ft = fm.beginTransaction();
        ft.replace(R.id.menu_fl_select, sessionLauncherFragment);
        ft.commit();
    }
}

File: app/src/main/java/com/example/mlallemant/mentalbattle/ui/menu/fragment/CreateJoinFragment.kt
package com.example.mlallemant.mentalbattle.ui.menu.fragment

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import com.example.mlallemant.mentalbattle.R
import com.example.mlallemant.mentalbattle.databinding.MenuCreateFragmentBinding
import com.example.mlallemant.mentalbattle.ui.extention.toast
import com.example.mlallemant.mentalbattle.ui.menu.MenuActivity
import com.example.mlallemant.mentalbattle.ui.menu.MenuActivity.Companion.BUNDLE_EXTRA_CREATOR
import com.example.mlallemant.mentalbattle.ui.menu.MenuActivity.Companion.BUNDLE_EXTRA_IS_CREATOR
import com.example.mlallemant.mentalbattle.ui.menu.MenuActivity.Companion.BUNDLE_EXTRA_NAME
import com.example.mlallemant.mentalbattle.ui.menu.MenuActivity.Companion.BUNDLE_EXTRA_PASSWORD
import com.example.mlallemant.mentalbattle.utils.DatabaseManager
import com.example.mlallemant.mentalbattle.utils.Player
import com.example.mlallemant.mentalbattle.utils.Session
import com.example.mlallemant.mentalbattle.utils.Utils

/**
 * Created by m.lallemant on 15/11/2017.
 */
class CreateJoinFragment : Fragment() {

    private var _binding: MenuCreateFragmentBinding? = null
    private val binding get() = _binding!!

    //Utils
    private var currentPlayer: Player? = null
    private lateinit var db: DatabaseManager
    private var isCreator = false

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        _binding = MenuCreateFragmentBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        val menuActivity = activity as MenuActivity?
        currentPlayer = menuActivity?.currentPlayer
        isCreator = arguments?.getBoolean(BUNDLE_EXTRA_CREATOR) ?: false
        db = DatabaseManager.getInstance()
        db.deletePlayerInLobby(currentPlayer)
        initUI()
        initListener()
    }

    private fun initUI() {
        with(binding) {
            pbCreateSession.visibility = View.INVISIBLE
            btnCreateSession.text = when (isCreator) {
                true -> getString(R.string.create)
                else -> getString(R.string.join)
            }
            tvTitleSession.text = when (isCreator) {
                true -> getString(R.string.create_session)
                else -> getString(R.string.join_session)
            }
        }
    }

    private fun initListener() {
        with(binding) {
            ivCreateBack.setOnClickListener {
                returnSelectorFragment()
                db.insertPlayerInLobby(currentPlayer)
            }
            btnCreateSession.setOnClickListener {
                pbCreateSession.visibility = View.VISIBLE
                if (isCreator) {
                    createSession()
                } else {
                    joinSession()
                }
            }
        }
    }

    private fun createSession() {
        val sessionName = binding.etCreateSessionName.text.toString()
        val sessionPassword = binding.etCreateSessionPassword.text.toString()
        if (sessionName.length > 4 && sessionPassword.length > 5) {
            val session = Session(sessionName, sessionPassword, Utils.SESSION_STATE_WAITING)
            db.initCheckSessionExist(session)
            db.setOnSessionExistListener { isExist ->
                if (isExist) {
                    toast(getString(R.string.session_already_exist))
                } else {
                    db.insertSession(session)
                    val playerForSession = Player(
                        currentPlayer?.id,
                        currentPlayer?.name,
                        currentPlayer?.score,
                        null,
                        null,
                        null,
                        Utils.SESSION_RDY_YES,
                        Utils.SESSION_CREATOR
                    )
                    db.insertPlayerInSession(session, playerForSession)
                    launchSessionLauncherFragment(true)
                }
                binding.pbCreateSession.visibility = View.INVISIBLE
            }
        } else {
            toast(getString(R.string.name_or_pswd_not_long_enough))
        }
    }

    private fun joinSession() {
        val sessionName = binding.etCreateSessionName.text.toString()
        val sessionPassword = binding.etCreateSessionPassword.text.toString()
        if (sessionName.isNotEmpty() && sessionPassword.isNotEmpty()) {
            val session = Session(sessionName, sessionPassword, Utils.SESSION_STATE_WAITING)
            db.initCheckSessionExist(session)
            db.setOnSessionExistListener { isExist ->
                if (isExist) {
                    val playerForSession = Player(
                        currentPlayer?.id,
                        currentPlayer?.name,
                        currentPlayer?.score,
                        null,
                        null,
                        null,
                        Utils.SESSION_RDY_KO,
                        Utils.SESSION_NEW_YES
                    )
                    db.insertPlayerInSession(session, playerForSession)
                    launchSessionLauncherFragment(false)
                } else {
                    toast(getString(R.string.bad_name_or_pswd))
                }
                binding.pbCreateSession.visibility = View.INVISIBLE
            }
        } else {
            toast(getString(R.string.bad_name_or_pswd))
        }
    }

    private fun returnSelectorFragment() {
        val selectorFragment = SelectorFragment()
        fragmentManager?.beginTransaction()?.replace(R.id.menu_fl_select, selectorFragment)
            ?.commit()
    }

    private fun launchSessionLauncherFragment(isCreator: Boolean) {
        val sessionLauncherFragment = SessionLauncherFragment()
        val bundle = Bundle().apply {
            putBoolean(BUNDLE_EXTRA_IS_CREATOR, isCreator)
            putString(BUNDLE_EXTRA_NAME, binding.etCreateSessionName.text.toString())
            putString(BUNDLE_EXTRA_PASSWORD, binding.etCreateSessionPassword.text.toString())
        }
        sessionLauncherFragment.arguments = bundle
        fragmentManager?.beginTransaction()?.replace(R.id.menu_fl_select, sessionLauncherFragment)
            ?.commit()
    }
}